### PR TITLE
Products: add actions to handle products

### PR DESCRIPTION
**Pull request by nicordev**

### CODE REVIEW

File: src/Controller/ProductController.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    /**
+     * @Get(
+     *     path = "/products/{id}",
+     *     name = "product_show",
+     *     requirements = {"id": "\d+"}
+     * )
+     */
+    public function getProductAction(Product $product)
+    {
+        $data = $this->serializer->serialize($product, "json");
```
Comment (olix21) [inline src/Controller/ProductController.php]: Those 4 lines could be simplified: `return $this->json($data);` thanks to Symfony ;)

Comment (nicordev) [reply src/Controller/ProductController.php]: Done, but now I have some \n in my json response on Postman

File: src/Controller/ProductController.php [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+     * @Post(
+     *     "/products/{id}",
+     *     name = "product_edit",
+     *     requirements = {"id": "\d+"}
+     * )
+     */
+    public function editAction(Request $request, Product $product, EntityManagerInterface $manager)
+    {
+        $data = $request->getContent();
+        $editedProduct = $this->serializer->deserialize($data, "App\\Entity\\Product", "json");
```
Comment (olix21) [inline src/Controller/ProductController.php]: Product::class

File: src/Controller/ProductController.php [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+class ProductController extends AbstractFOSRestController
+{
+    /**
+     * @var SerializerInterface
+     */
+    private $serializer;
+
+    public function __construct(SerializerInterface $serializer)
+    {
+        $this->serializer = $serializer;
```
Comment (olix21) [inline src/Controller/ProductController.php]: seems like your injecting something you don't use

File: src/Controller/ProductController.php [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    /**
+     * @Get(
+     *     path = "/products",
+     *     name = "product_list"
+     * )
+     * @Rest\QueryParam(
+     *     name = "property",
+     *     requirements = "id|price|quantity|brand|model",
+     *     default = "price",
+     *     description = "Sort order (asc or desc)"
```
Comment (olix21) [inline src/Controller/ProductController.php]: wrong description

File: src/Controller/ProductController.php [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    ) {
+        if ($search !== null) {
+            if (in_array($property, ["brand", "model"])) {
+                $criteria = [$property => $search];
+            } else {
+                return new Response("Can not use search parameter, the property is either missing or wrong.", Response::HTTP_NOT_ACCEPTABLE);
+            }
+        }
+
+        $exact === "false" ? $exactValue = false : $exactValue = true;
```
Comment (olix21) [inline src/Controller/ProductController.php]: `$exactValue = $exact !== "false";`

File: src/Controller/ProductController.php (hunk 2 above)
Comment (nicordev) [reply src/Controller/ProductController.php]: Already done in later commits.

File: src/Controller/ProductController.php (hunk 5 above)
Comment (nicordev) [reply src/Controller/ProductController.php]: Refactored. But what would be the best way to handle this?

I'm using a string here, but at first, I used a parameter `bool $exact` so the query was written like that: `exact=0` for false and `exact=1` for true.

Which one should I use? Or maybe there is a better way than these two?

File: src/EventListener/ExceptionListener.php [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+namespace App\EventListener;
+
+
+use App\Exception\ResourceValidationException;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Event\ExceptionEvent;
+
+class ExceptionListener
+{
+    public function onKernelException(ExceptionEvent $event)
```
Comment (olix21) [inline src/EventListener/ExceptionListener.php]: Could be avoided if you use the fos exception controller:  https://symfony.com/doc/master/bundles/FOSRestBundle/4-exception-controller-support.html

**Review by olix21 — APPROVED** [review]

File: src/EventListener/ExceptionListener.php (hunk 6 above)
Comment (nicordev) [reply src/EventListener/ExceptionListener.php]: Done using [these configuration settings](https://openclassrooms.com/fr/courses/4087036-construisez-une-api-rest-avec-symfony/4333906-gestion-des-erreurs#/id/r-4336864)

**Pull request merged**